listenerutil/forwarded_for.go

```diff
@@ -0,0 +1,192 @@
+// encounters an error
+type ErrResponseFn func(w http.ResponseWriter, status int, err error)
+
+// WrapForwaredForHandler is an http middleware handler which uses the
+// XForwardedFor* listener config settings to determine how/if X-Forwarded-For
+// are trusted/allowed for an inbound request.  In the end, if a "trusted"
+// X-Forwarded-For header is found, then the request RemoteAddr will be
+// overwritten with it before the request is served.
+func WrapForwardedForHandler(h http.Handler, l *ListenerConfig, respErrFn ErrResponseFn) (http.Handler, error) {
+	if h == nil {
+		return nil, fmt.Errorf("missing http handler: %w", ErrInvalidParameter)
+	}
+	if l == nil {
+		return nil, fmt.Errorf("missing listener config: %w", ErrInvalidParameter)
+	}
+	if respErrFn == nil {
+		return nil, fmt.Errorf("missing response error function: %w", ErrInvalidParameter)
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		trusted, err := TrustedFromXForwardedFor(r, l)
+		if err != nil {
+			respErrFn(w, http.StatusBadRequest, err)
+			return
+		}
+		if trusted == nil {
+			h.ServeHTTP(w, r)
+			return
+		}
+		newCtx, err := newOrigRemoteAddrCtx(r.Context(), r.RemoteAddr)
+		if err != nil {
+			respErrFn(w, http.StatusBadRequest, fmt.Errorf("error setting orig remote header ctx: %w", err))
+			return
+		}
+		r = r.WithContext(newCtx)
+		r.RemoteAddr = net.JoinHostPort(trusted.Host, trusted.Port)
+		h.ServeHTTP(w, r)
+		return
+	}), nil
+}
+
+// Addr represents only the Host and Port of a TCP address.
+type Addr struct {
+	Host string
+	Port string
+}
+
+// TrustedFromXForwardedFor will use the XForwardedFor* listener config settings
+// to determine how/if X-Forwarded-For are trusted/allowed for an inbound
+// request.  Important: return values of nil, nil are valid and simply means
+// that no "trusted" header was found and no error was raised as well.  Errors
+// can be raised for a number of conditions based on the listener config
+// settings, especially when the config setting for
+// XForwardedForRejectNotPresent is set to true which means if a "trusted"
+// header can't be found the request should be rejected.
+func TrustedFromXForwardedFor(r *http.Request, l *ListenerConfig) (*Addr, error) {
+	if r == nil {
+		return nil, fmt.Errorf("missing http request: %w", ErrInvalidParameter)
+	}
+	if l == nil {
+		return nil, fmt.Errorf("missing listener config: %w", ErrInvalidParameter)
+	}
+	rejectNotPresent := l.XForwardedForRejectNotPresent
+	hopSkips := l.XForwardedForHopSkips
+	authorizedAddrs := l.XForwardedForAuthorizedAddrs
+	rejectNotAuthz := l.XForwardedForRejectNotAuthorized
+
+	headers, headersOK := r.Header[textproto.CanonicalMIMEHeaderKey("X-Forwarded-For")]
+	if !headersOK || len(headers) == 0 {
+		if !rejectNotPresent {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("missing x-forwarded-for header and configured to reject when not present")
+	}
+
+	host, port, err := net.SplitHostPort(r.RemoteAddr)
```

Comment:
I'm not sure if RemoteAddr will always contain a port; if not, be aware that SplitHostPort errors if there is no port segment, so you may want to account for this.

Reply:
I agree that it won't always contain a port.  In this case, an error will be raised and handled based on the value of `rejectNotPresent`.  
This mimics how vault handles this case: https://github.com/hashicorp/vault/blob/87c355e2bb34018b03e886f2e9155a6b3aba69fb/http/handler.go#L488
Do we want to change how it's handled?

Reply:
Yes, I think so. Looking at MDN: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For in all of their examples no port is present. Even the examples for the standardized version of it (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded) don't have ports shown.

Probably Vault has this wrong. @briankassouf curious your thoughts around this.

Reply:
This is parsing the RemoteAddr that is provided in the http.Request object, which, may always have a port? https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/net/http/request.go;l=279-286

Reply:
I agree X-Forwarded-For headers may not have a port, but i don't think we have that invariant below when parsing those values

Reply:
Oh, I'm sorry. In this re-review I mixed up RemoteAddr vs X-Forwarded-For as you can clearly see in my above comments :-)